packages/core/src/subscriptions/index.ts

```diff
@@ -0,0 +1,259 @@
+import { Bundle, Parameters, Subscription, SubscriptionStatus } from '@medplum/fhirtypes';
+import { MedplumClient } from '../client';
+import { TypedEventTarget } from '../eventtarget';
+import { OperationOutcomeError, badRequest, serverError, validationError } from '../outcomes';
+import { ProfileResource, getReferenceString, resolveId } from '../utils';
+
+export type SubscriptionEventMap = {
+  connect: { type: 'connect'; payload: { subscriptionId: string } };
+  disconnect: { type: 'disconnect'; payload: { subscriptionId: string } };
+  error: { type: 'error'; payload: Error };
+  message: { type: 'message'; payload: Bundle };
+  close: { type: 'close' };
+  heartbeat: { type: 'heartbeat'; payload: Bundle };
+};
+
+const kAddCriteria = Symbol.for('medplum.SubscriptionEmitter.addCriteria');
+const kRemoveCriteria = Symbol.for('medplum.SubscriptionEmitter.removeCriteria');
+
+/**
+ * An `EventTarget` that emits events when new subscription notifications come in over WebSockets.
+ *
+ * -----
+ *
+ * ### Events emitted:
+ *
+ * - `connect` - A new subscription is connected to the `SubscriptionManager` and `message` events for this subscription can be expected.
+ * - `disconnect` - The specified subscription is no longer being monitored by the `SubscriptionManager`.
+ * - `error` - An error has occurred.
+ * - `message` - A message containing a notification `Bundle` has been received.
+ * - `close` - The WebSocket has been closed.
+ * - `heartbeat` - A `heartbeat` message has been received.
+ */
+export class SubscriptionEmitter extends TypedEventTarget<SubscriptionEventMap> {
+  private criteria: Set<string>;
+  constructor(...criteria: string[]) {
+    super();
+    this.criteria = new Set(criteria);
+  }
+  getCriteria(): Set<string> {
+    return this.criteria;
+  }
+  [kAddCriteria](criteria: string): void {
+    this.criteria.add(criteria);
+  }
+  [kRemoveCriteria](criteria: string): void {
+    this.criteria.delete(criteria);
+  }
+}
+
+export class SubscriptionManager {
+  private readonly medplum: MedplumClient;
+  private ws: WebSocket;
+  private masterSubEmitter?: SubscriptionEmitter;
+  private subEmitters: Map<string, SubscriptionEmitter>; // Map<criteria, SubscriptionEmitter>
+  private refCounts: Map<string, number>; // Map<criteria, refCount>
+  private subscriptionCriteriaLookup: Map<string, string>; // Map<subscriptionId, criteria>
+  private criteriaSubscriptionLookup: Map<string, string>; // Map<criteria, subscriptionId>
```

Comment:
Rather than using `Map` for everything, I wonder if we could add a helper class or type which would contain the logic.  It could be quite small, be private/internal, and manage this state to avoid the lookups.

```ts
class SubscriptionCriteria {
  readonly criteria: string;
  readonly subscriptionId: string;
  readonly emitter: SubscriptionEmitter;
  refCount: number;
}
```

Reply:
We still will need the lookups, there is no other way to associate `criteria` and `subscriptionId` since these exist separately and you cannot get the `criteria` for `subscription` without the lookup and vice versa.

Reply:
The `subscription -> criteria` is needed since the `SubscriptionStatus` only contains a reference string and not the full resource as a convention. So we need to cache the associated `criteria` so we know which callback to call when a notification comes in

Reply:
The class might still be useful though